test.pl: Use END blocks to cancel watchdog timers

khwilliamson opened the pull request:
In doing follow up work to a7f3f23
"Turn off watchdog when done in tests", I realized that my solution was
suboptimal, and that the code already existed in test.pl to do things
better.

What that code does is to create an END block to cancel the watchdog
upon program exit.  I think that is a better solution than to force
the addition of explicit calls to watchdog(0).  This commit creates a
standard way to specify the code that does the cancellation, and to use
it, not only when the END block does, but also when the timer is
explicitly cancelled.

This means the calls to watchdog(0) that occur at the end of the file
that were added in a7f3f23 aren't
necessary.  The END block takes care of it.  The reason to keep them is
if we want to add a porting test that every watchdog is cleared.

* This set of changes does not require a perldelta entry.

khwilliamson committed:
Remove explicit watchdog cancellations at end of program

The previous commit made these unnecessary.  This commit removes them.
I also removed a couple that occurred just before done_testing() and
plan() that occur at the end of the program.

tonycoz commented on t/test.pl:
```diff
@@ -2022,6 +2009,9 @@ ($;$)
                 kill($sig, $pid_to_kill);
         });
 
+        $cancel_string =~ s/HERE/$watchdog_thread->kill('KILL')->detach()/;
```
`$watchdog_thread` would be replaced at this point, producing code like `threads=SCALAR(0x55e3bd6a7ab0)->kill('KILL')->detach()` which will just fail to compile.

khwilliamson replied:
This was wishful thinking on my part

tonycoz commented on t/test.pl:
```diff
@@ -2027,12 +2013,15 @@ ($;$)
                 kill($sig, $pid_to_kill);
         });
 
+        $cancel_string =~ s/HERE/$::watchdog_thread_->kill('KILL')->detach()/;
```
Still gets replaced badly:
```
$ perl -Mthreads -E '$cancel_string = "END{  HERE }"; our $watchdog_thread_ = threads->create(sub {}); $watchdog_thread_->join; $cancel_string =~ s/HERE/$::watchdog_thread_->kill('KILL')->detach()/; say $cancel_string'
END{  threads=SCALAR(0x562c8311de88)->kill(KILL)->detach() }
```
You need to escape the `$` in the s/// replacement string.

khwilliamson replied:
its interesting that nothing fails.

khwilliamson committed:
test.pl: Clarify some comments

khwilliamson committed:
test.pl: Don't use watchdog thread unless safe signals

In the thread documentation, I noticed that unsafe signals preclude
killing threads with a signal.  But this is the normal method for
clearing a watchdog timer under threads.  I doubt that people are
compiling with PERL_OLD_SIGNALS these days, so I didn't check for that,
but its easy enough to check for the environment variable that does the
same thing at runtime.

khwilliamson committed:
test.pl: Use END blocks to cancel watchdog timers

In doing follow up work to a7f3f23c846f4fd664a718235fdb2ad97992c26e
"Turn off watchdog when done in tests", I realized that my solution was
suboptimal, and that the code already existed in test.pl to do things
better.

What that code does is to create an END block to cancel the watchdog
upon program exit.  I think that is a better solution than to force
the addition of explicit calls to watchdog(0).  This commit creates a
standard way to specify the code that does the cancellation, and to use
it, not only when the END block does, but also when the timer is
explicitly cancelled.  That standard method is to create a string that
basically gets evaled.   This entailed making any watchdog thread object
into a global so that a reference to it could be stringified for the
eval.

This means the calls to watchdog(0) that occur at the end of the file
that were added in a7f3f23c846f4fd664a718235fdb2ad97992c26e aren't
necessary.  The END block takes care of it.  The reason to keep them
would be if we want to add a porting test that every watchdog is cleared.

khwilliamson commented:
In looking at things more closely, I discovered the possibility of unsafe signals, so I added a check against that

I also changed so with threads you can have multiple alarms going

tonycoz commented:
> In looking at things more closely, I discovered the possibility of unsafe signals, so I added a check against that

I'll admit I was confused by this, since KILL isn't catchable, but threads uses the safe signal delivery mechanism.

But the watchdog(0) signal still doesn't trigger the KILL handler set in the watchdog thread and I don't know why (yet anyway).

tonycoz commented on t/test.pl:
```diff
@@ -2029,12 +2020,16 @@ ($;$)
                 kill($sig, $pid_to_kill);
         });
 
+        #my $index = scalar @watchd
```
comment here seems leftover

tonycoz commented:
Is the intent here to support nested watchdog timers?  I don't see much other reason to mess around with the evalled END blocks here.

If you didn't want nested watchdogs the original code could have worked, with an `END{ watchdog(0) }` to ensure any pending watchdogs were cleared.
